src/logical/model.rs

```diff
@@ -371,19 +371,41 @@ impl Filter {
 pub struct Rule {
     /// Head atoms of the rule
     head: Vec<Atom>,
-    /// Body literals of the rule
-    body: Vec<Literal>,
+    /// Positive Body literals of the rule
+    positive_body: Vec<Atom>,
     /// Filters applied to the body
-    filters: Vec<Filter>,
+    positive_filters: Vec<Filter>,
+    /// Negative Body literals of the rule
+    negative_body: Vec<Atom>,
+    /// Filters applied to the body
+    negative_filters: Vec<Filter>,
 }
 
 impl Rule {
+    fn split_literal_vector(literals: Vec<Literal>) -> (Vec<Atom>, Vec<Atom>) {
+        let mut positive = Vec::new();
+        let mut negative = Vec::new();
+
+        for literal in literals {
+            match literal {
+                Literal::Positive(atom) => positive.push(atom),
+                Literal::Negative(atom) => negative.push(atom),
+            }
+        }
+
+        (positive, negative)
+    }
+
     /// Construct a new rule.
-    pub fn new(head: Vec<Atom>, body: Vec<Literal>, filters: Vec<Filter>) -> Self {
+    pub fn new(head: Vec<Atom>, body: Vec<Literal>, positive_filters: Vec<Filter>) -> Self {
+        let (positive_body, negative_body) = Self::split_literal_vector(body);
```

Comment:
I find it odd to have `split_literal_vector` as associated method on `Rule` when it is really independent of rules.
Maybe just inline what the method does right here.
Also, did you think about using [`partition`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition)?

Reply:
Partition is not that convenient since I get two vectors of literals instead of two vectors of atoms. I went with the inline solution.

Reply:
I see. Looks fine :)